File: trees/binary-tree-inorder/binary-tree-inorder.py
"""Trees: Inorder traversal of a binary tree"""

from __future__ import annotations


class TreeNode:
    def __init__(self, val: int, left: "TreeNode | None" = None, right: "TreeNode | None" = None):
        self.val = val
        self.left = left
        self.right = right


def inorder(root: TreeNode | None, result: list[int]) -> None:
    if root is None:
        return
    inorder(root.left, result)
    result.append(root.val)
    inorder(root.right, result)


if __name__ == "__main__":
    root = TreeNode(2, TreeNode(1), TreeNode(3))

    result: list[int] = []
    inorder(root, result)
    print(result)

File: trees/binary-tree-inorder/binary-tree-inorder.cpp
// Trees: Inorder traversal of a binary tree
#include <iostream>
#include <vector>
using namespace std;

struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int v) : val(v), left(nullptr), right(nullptr) {}
};

void inorder(TreeNode* root, vector<int>& result) {
    if (root == nullptr) return;
    inorder(root->left, result);
    result.push_back(root->val);
    inorder(root->right, result);
}

int main() {
    TreeNode* root = new TreeNode(2);
    root->left = new TreeNode(1);
    root->right = new TreeNode(3);

    vector<int> result;
    inorder(root, result);

    for (int val : result) cout << val << " ";
    cout << endl;

    return 0;
}
